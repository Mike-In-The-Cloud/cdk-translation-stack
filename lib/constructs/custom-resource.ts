import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as custom_resources from 'aws-cdk-lib/custom-resources';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';

export interface CustomResourceConstructProps {
  readonly targetLambdaFunction: lambda.IFunction;
  readonly stateMachine: stepfunctions.IStateMachine;
  readonly customResourceProviderLambda: lambda.IFunction; // The lambda that implements the CR logic
}

export class CustomResourceConstruct extends Construct {
  constructor(scope: Construct, id: string, props: CustomResourceConstructProps) {
    super(scope, id);

    // Policy for the CR Provider Lambda to update the target Lambda's configuration
    const updateLambdaConfigPolicy = new iam.PolicyStatement({
      actions: ['lambda:UpdateFunctionConfiguration'],
      resources: [props.targetLambdaFunction.functionArn],
    });

    if (props.customResourceProviderLambda.role) {
        props.customResourceProviderLambda.role.addToPrincipalPolicy(updateLambdaConfigPolicy);
    } else {
        // If the role is not available on the IFunction (e.g. imported function),
        // this might indicate a need for a different approach or more specific role handling.
        // For now, we proceed assuming the role is typically available for functions created in the stack.
        console.warn(`CustomResourceProviderLambda (${props.customResourceProviderLambda.functionName}) has no role attached. Cannot add updateLambdaConfigPolicy.`);
    }

    // Custom Resource properties
    const customResourceProps = {
      FunctionName: props.targetLambdaFunction.functionName,
      EnvironmentVariables: {
        TRANSLATION_STATE_MACHINE_ARN: props.stateMachine.stateMachineArn,
      },
    };

    // Custom Resource Provider
    const provider = new custom_resources.Provider(this, 'SetEnvVarCrProvider', {
      onEventHandler: props.customResourceProviderLambda,
    });

    // Custom Resource
    const setEnvVarCustomResource = new cdk.CustomResource(this, 'SetLambdaEnvVarCr', {
      serviceToken: provider.serviceToken,
      properties: customResourceProps,
    });

    // Dependencies for the Custom Resource
    // Ensures the target lambda exists before attempting to update it.
    setEnvVarCustomResource.node.addDependency(props.targetLambdaFunction);
    // Ensures the state machine exists, so its ARN is available.
    setEnvVarCustomResource.node.addDependency(props.stateMachine);
    // Ensures the CR provider lambda and its role (if applicable) are created before the CR tries to use it.
    setEnvVarCustomResource.node.addDependency(props.customResourceProviderLambda);
    if (props.customResourceProviderLambda.role) {
        setEnvVarCustomResource.node.addDependency(props.customResourceProviderLambda.role);
    }
    setEnvVarCustomResource.node.addDependency(provider); // Dependency on the provider itself
  }
}