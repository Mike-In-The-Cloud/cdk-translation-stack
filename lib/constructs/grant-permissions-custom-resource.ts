import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as custom_resources from 'aws-cdk-lib/custom-resources';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';

export interface GrantPermissionsCustomResourceProps {
  readonly lambdaToGrantStartExecution: lambda.IFunction; // e.g., DocumentProcessor
  readonly stateMachineToGrantInvoke: stepfunctions.IStateMachine;
  readonly stateMachineRoleToGrantInvoke: iam.IRole;
  readonly lambdasToInvokeByStateMachine: lambda.IFunction[];
  readonly customResourceProviderLambda: lambda.IFunction; // The grantPermissions lambda
}

export class GrantPermissionsCustomResource extends Construct {
  constructor(scope: Construct, id: string, props: GrantPermissionsCustomResourceProps) {
    super(scope, id);

    // Custom Resource properties to pass to the grantPermissionsLambda
    const customResourceProps = {
      LambdaRoleArn: props.lambdaToGrantStartExecution.role!.roleArn, // Role of the Lambda needing StartExecution
      StateMachineRoleArn: props.stateMachineRoleToGrantInvoke.roleArn, // Role of the State Machine needing InvokeFunction
      StateMachineArn: props.stateMachineToGrantInvoke.stateMachineArn,
      TargetLambdaArns: props.lambdasToInvokeByStateMachine.map(fn => fn.functionArn),
    };

    // Custom Resource Provider
    const provider = new custom_resources.Provider(this, 'GrantPermsCrProvider', {
      onEventHandler: props.customResourceProviderLambda,
      // Ensure the log group for the provider lambda exists or can be created.
      // By default, CDK creates a log group. If managing manually, ensure it has a retention policy.
    });

    // Custom Resource
    const grantPermsCustomResource = new cdk.CustomResource(this, 'GrantPermsCr', {
      serviceToken: provider.serviceToken,
      properties: customResourceProps,
    });

    // Dependencies for the Custom Resource
    // Ensures the roles and functions exist before trying to update their policies.
    grantPermsCustomResource.node.addDependency(props.lambdaToGrantStartExecution.role!);
    grantPermsCustomResource.node.addDependency(props.stateMachineRoleToGrantInvoke);
    grantPermsCustomResource.node.addDependency(props.stateMachineToGrantInvoke);
    props.lambdasToInvokeByStateMachine.forEach(fn => grantPermsCustomResource.node.addDependency(fn));

    // Ensures the CR provider lambda and its role are created before the CR tries to use it.
    grantPermsCustomResource.node.addDependency(props.customResourceProviderLambda);
    if (props.customResourceProviderLambda.role) {
        grantPermsCustomResource.node.addDependency(props.customResourceProviderLambda.role);
    }
    grantPermsCustomResource.node.addDependency(provider);
  }
}