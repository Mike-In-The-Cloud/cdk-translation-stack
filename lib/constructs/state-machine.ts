import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import * as lambda from 'aws-cdk-lib/aws-lambda'; // Import lambda types
import * as iam from 'aws-cdk-lib/aws-iam'; // Import iam types

export interface StateMachineConstructProps {
  // Accept individual functions instead of the whole construct
  readonly documentProcessorLambda: lambda.IFunction;
  readonly translationProcessorLambda: lambda.IFunction;
  readonly documentCombinerLambda: lambda.IFunction;
  readonly notificationSenderLambda?: lambda.IFunction; // Optional
  // REMOVE: readonly role?: iam.IRole; // Role is now handled externally
}

export class StateMachineConstruct extends Construct {
  // Expose the definition, not the state machine resource
  public readonly definition: stepfunctions.IChainable;
  // REMOVE: public readonly stateMachine: stepfunctions.StateMachine;

  constructor(scope: Construct, id: string, props: StateMachineConstructProps) {
    super(scope, id);

    // Use the specific lambda props
    const {
      documentProcessorLambda,
      translationProcessorLambda,
      documentCombinerLambda,
      notificationSenderLambda
    } = props;

    // --- Step Functions workflow definition ---
    const validateDocument = new tasks.LambdaInvoke(this, 'ValidateDocument', {
      lambdaFunction: documentProcessorLambda, // Use prop
      outputPath: '$.Payload',
    });

    const translateSection = new tasks.LambdaInvoke(this, 'TranslateSection', {
      lambdaFunction: translationProcessorLambda, // Use prop
      inputPath: '$',
      outputPath: '$.Payload',
    });

    const combineDocuments = new tasks.LambdaInvoke(this, 'CombineDocuments', {
      lambdaFunction: documentCombinerLambda, // Use prop
      inputPath: '$',
      outputPath: '$.Payload',
    });

    // Map state to process sections in parallel
    const mapState = new stepfunctions.Map(this, 'MapState', {
        maxConcurrency: 5, // Example concurrency limit
        itemsPath: stepfunctions.JsonPath.stringAt('$.documentSections'), // Assumes validateDocument outputs { ..., documentSections: [...] }
        inputPath: '$', // Pass the full input to the Map state iterations
        parameters: { // Define input for each iteration (translateSection lambda)
            "bucket.$'": '$.bucket', // Pass bucket from original input
            "key.$'": '$.key', // Pass key from original input
            "sourceLanguage.$'": '$.sourceLanguage', // Pass sourceLanguage from original input
            "targetLanguage.$'": '$.targetLanguage', // Pass targetLanguage from original input
            "section.$'": stepfunctions.JsonPath.stringAt('$$.Map.Item.Value') // Pass the current item from itemsPath
        },
        resultPath: '$.translatedSections' // Collect results under this key
    });
    mapState.iterator(translateSection); // Define the task to run for each item


    let workflowDefinition = stepfunctions.Chain.start(validateDocument)
      .next(mapState) // Use Map state instead of Parallel
      .next(combineDocuments);

    // --- Handle Optional Notification Sender in Workflow ---
    if (notificationSenderLambda) { // Check the specific optional prop
      const sendNotification = new tasks.LambdaInvoke(this, 'SendNotification', {
        lambdaFunction: notificationSenderLambda, // Use prop
        inputPath: '$', // Pass the output of combineDocuments
      });
      // Chain notification after combineDocuments
      // Make sure combineDocuments returns necessary info for notification
      workflowDefinition = workflowDefinition.next(sendNotification);
    }

    // Assign the final definition chain to the public property
    this.definition = workflowDefinition;

    // REMOVE: State Machine creation logic
    // const stateMachineName = `${cdk.Aws.STACK_NAME}-TranslationWorkflow`;
    // this.stateMachine = new stepfunctions.StateMachine(this, 'TranslationStateMachine', {
    //   stateMachineName: stateMachineName,
    //   definitionBody: stepfunctions.DefinitionBody.fromChainable(workflowDefinition),
    //   role: props.role, // Use passed role or let it create one
    // });
  }
}