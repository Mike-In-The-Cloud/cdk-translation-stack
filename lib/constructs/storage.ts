import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as opensearch from 'aws-cdk-lib/aws-opensearchserverless';
import { Construct } from 'constructs';

export interface StorageConstructProps {
  tmxBucketName: string;
  documentsBucketName: string;
  openSearchCollectionName: string;
  // Add other relevant config like removal policies if needed
}

export class StorageConstruct extends Construct {
  public readonly tmxBucket: s3.Bucket;
  public readonly documentsBucket: s3.Bucket;
  public readonly translationCollection: opensearch.CfnCollection;

  constructor(scope: Construct, id: string, props: StorageConstructProps) {
    super(scope, id);

    this.tmxBucket = new s3.Bucket(this, 'TMXBucket', {
      bucketName: props.tmxBucketName,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Consider making this configurable
    });

    this.documentsBucket = new s3.Bucket(this, 'DocumentsBucket', {
      bucketName: props.documentsBucketName,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Consider making this configurable
    });

    this.translationCollection = new opensearch.CfnCollection(this, 'TranslationCollection', {
      name: props.openSearchCollectionName,
      description: 'Collection for storing translation embeddings',
      type: 'VECTORSEARCH'
    });
  }
}