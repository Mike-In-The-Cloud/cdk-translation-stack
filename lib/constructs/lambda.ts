import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as nodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import * as path from 'path';
import { TranslationStackConfig } from '../config'; // Assuming config types are here
import { StorageConstruct } from './storage'; // Need storage outputs

export interface LambdaConstructProps {
    config: TranslationStackConfig;
    storage: StorageConstruct;
    // REMOVE: readonly stateMachineArn: string;
}

export class LambdaConstruct extends Construct {
    // Expose functions needed by other constructs/stack
    public readonly tmxProcessor: nodejs.NodejsFunction;
    public readonly documentProcessor: nodejs.NodejsFunction;
    public readonly translationProcessor: nodejs.NodejsFunction;
    public readonly documentCombiner: nodejs.NodejsFunction;
    public readonly setLambdaEnvVarFunction: nodejs.NodejsFunction;
    public readonly notificationSender?: nodejs.NodejsFunction; // Optional
    public readonly grantPermissionsFunction: nodejs.NodejsFunction; // Function for GrantPermissions CR

    constructor(scope: Construct, id: string, props: LambdaConstructProps) {
        super(scope, id);

        // REMOVE: const { config, storage, stateMachineArn } = props;
        const { config, storage } = props; // stateMachineArn is no longer a prop
        const region = cdk.Stack.of(this).region;
        const account = cdk.Stack.of(this).account;

        // Common NodejsFunction settings
        const nodeJsFunctionProps: Partial<nodejs.NodejsFunctionProps> = {
            runtime: lambda.Runtime.NODEJS_18_X,
            handler: 'handler',
            bundling: {
                externalModules: [ // Optimize bundling if needed
                    '@aws-sdk/*' // Exclude AWS SDK v3 clients if using Lambda Layer or system provided
                ],
                nodeModules: ['@aws-sdk/client-iam'],
            },
            timeout: cdk.Duration.minutes(1),
        };

        // --- Function Definitions ---

        this.tmxProcessor = new nodejs.NodejsFunction(this, 'TMXProcessor', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/functions/tmxProcessor.ts'),
            timeout: cdk.Duration.seconds(30),
            environment: {
                OPENSEARCH_ENDPOINT: storage.translationCollection.attrCollectionEndpoint,
                BEDROCK_EMBEDDING_MODEL: config.bedrockConfig.embeddingModel,
            },
            initialPolicy: [
                new iam.PolicyStatement({
                    actions: ['bedrock:InvokeModel'],
                    resources: [`arn:aws:bedrock:${region}:${account}:model/${config.bedrockConfig.embeddingModel}`],
                }),
            ],
        });

        this.documentProcessor = new nodejs.NodejsFunction(this, 'DocumentProcessor', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/functions/documentProcessor.ts'),
            timeout: cdk.Duration.seconds(30),
        });

        this.translationProcessor = new nodejs.NodejsFunction(this, 'TranslationProcessor', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/functions/translationProcessor.ts'),
            environment: {
                OPENSEARCH_ENDPOINT: storage.translationCollection.attrCollectionEndpoint,
                BEDROCK_EMBEDDING_MODEL: config.bedrockConfig.embeddingModel,
                BEDROCK_TRANSLATION_MODEL: config.bedrockConfig.translationModel,
            },
            initialPolicy: [
                new iam.PolicyStatement({
                    actions: ['bedrock:InvokeModel'],
                    resources: [
                        `arn:aws:bedrock:${region}:${account}:model/${config.bedrockConfig.embeddingModel}`,
                        `arn:aws:bedrock:${region}:${account}:model/${config.bedrockConfig.translationModel}`,
                    ],
                }),
                // OpenSearch permission is granted via CfnAccessPolicy in the main stack
            ],
            timeout: cdk.Duration.seconds(30),
        });

        this.documentCombiner = new nodejs.NodejsFunction(this, 'DocumentCombiner', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/functions/documentCombiner.ts'),
            timeout: cdk.Duration.seconds(30),
        });

        // Custom Resource Lambda
        this.setLambdaEnvVarFunction = new nodejs.NodejsFunction(this, 'SetLambdaEnvVarFunction', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/custom/setLambdaEnvVar.ts'),
        });

        // Optional Notification Sender
        if (config.features?.emailNotifications && config.senderEmail) {
            this.notificationSender = new nodejs.NodejsFunction(this, 'NotificationSender', {
                ...nodeJsFunctionProps,
                entry: path.join(__dirname, '../../lambda/functions/notificationSender.ts'),
                environment: {
                    SENDER_EMAIL: config.senderEmail,
                },
                initialPolicy: [
                    new iam.PolicyStatement({
                        actions: ['ses:SendEmail', 'ses:SendRawEmail'],
                        resources: [`arn:aws:ses:${region}:${account}:identity/${config.senderEmail}`],
                    }),
                ],
                timeout: cdk.Duration.seconds(30),
            });
        }

        // Custom Resource Lambda (Grant Permissions)
        this.grantPermissionsFunction = new nodejs.NodejsFunction(this, 'GrantPermissionsFunction', {
            ...nodeJsFunctionProps,
            entry: path.join(__dirname, '../../lambda/custom/grantPermissions.ts'),
            description: 'Custom Resource handler to grant IAM permissions post-creation',
            initialPolicy: [
                // This Lambda needs permission to modify the policies of the target roles
                new iam.PolicyStatement({
                    actions: ['iam:PutRolePolicy', 'iam:DeleteRolePolicy'],
                    // WARNING: Scoping this is crucial for security.
                    // Ideally, scope this ONLY to the specific roles being modified.
                    // However, the role ARNs aren't known here yet.
                    // Using a wildcard is less secure but might be necessary if exact ARNs
                    // cannot be easily passed or derived during synth.
                    // Consider passing Role Names via environment variables if possible for tighter scope.
                    resources: ["*"] // TODO: Refine this resource scope if possible!
                }),
            ],
        });
    }
}