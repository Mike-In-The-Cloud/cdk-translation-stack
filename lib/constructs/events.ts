import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import { LambdaConstruct } from './lambda';
import { StorageConstruct } from './storage';

export interface EventsConstructProps {
  lambdaFunctions: LambdaConstruct;
  storage: StorageConstruct;
}

export class EventsConstruct extends Construct {
  constructor(scope: Construct, id: string, props: EventsConstructProps) {
    super(scope, id);

    const { lambdaFunctions, storage } = props;

    // --- EventBridge Rules ---
    const tmxRule = new events.Rule(this, 'TMXUploadRule', {
      // ruleName: `${cdk.Aws.STACK_NAME}-TMXUploadRule`, // Optional: Define a specific name
      description: 'Rule to trigger TMX processing lambda on S3 TMX bucket uploads',
      eventPattern: {
        source: ['aws.s3'],
        detailType: ['AWS API Call via CloudTrail'], // Using CloudTrail events for more reliable PutObject/CompleteMultipartUpload
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject', 'CompleteMultipartUpload'],
          requestParameters: {
            bucketName: [storage.tmxBucket.bucketName]
          }
        }
      }
    });

    const documentRule = new events.Rule(this, 'DocumentUploadRule', {
      // ruleName: `${cdk.Aws.STACK_NAME}-DocumentUploadRule`, // Optional: Define a specific name
      description: 'Rule to trigger document processing lambda on S3 documents bucket uploads',
      eventPattern: {
        source: ['aws.s3'],
        detailType: ['AWS API Call via CloudTrail'],
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject', 'CompleteMultipartUpload'],
          requestParameters: {
            bucketName: [storage.documentsBucket.bucketName]
          }
        }
      }
    });

    // Add EventBridge targets
    tmxRule.addTarget(new targets.LambdaFunction(lambdaFunctions.tmxProcessor));
    documentRule.addTarget(new targets.LambdaFunction(lambdaFunctions.documentProcessor));

    // Add permissions if needed (Lambda construct should ideally grant invoke permission, but double-check)
    // lambdaFunctions.tmxProcessor.addPermission(...)
    // lambdaFunctions.documentProcessor.addPermission(...)
  }
}