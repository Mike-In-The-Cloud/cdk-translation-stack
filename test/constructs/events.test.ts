import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { Construct } from 'constructs';
import { EventsConstruct } from '../../lib/constructs/events';
import { LambdaConstruct } from '../../lib/constructs/lambda';
import { StorageConstruct } from '../../lib/constructs/storage';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as nodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as fs from 'fs';

// Reusable helper for dummy NodejsFunction
function createDummyNodejsFunction(scope: Construct, id: string): nodejs.NodejsFunction {
    // Ensure dummy entry file exists
    if (!fs.existsSync('dummy-entry.ts')) {
        fs.writeFileSync('dummy-entry.ts', 'exports.handler = () => {};');
    }
    return new nodejs.NodejsFunction(scope, id, {
        runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
        handler: 'index.handler',
        entry: 'dummy-entry.ts',
        functionName: `dummy-${id}-func`
    });
}

// Mock LambdaConstruct
class MockLambdaConstruct extends Construct implements Pick<LambdaConstruct, 'tmxProcessor' | 'documentProcessor'> {
    public readonly tmxProcessor: nodejs.NodejsFunction;
    public readonly documentProcessor: nodejs.NodejsFunction;

    constructor(scope: Construct, id: string) {
        super(scope, id);
        this.tmxProcessor = createDummyNodejsFunction(this, 'TmxProcessor');
        this.documentProcessor = createDummyNodejsFunction(this, 'DocProcessor');
    }
}

// Mock StorageConstruct
class MockStorageConstruct extends Construct implements Pick<StorageConstruct, 'tmxBucket' | 'documentsBucket'> {
    public readonly tmxBucket: s3.Bucket;
    public readonly documentsBucket: s3.Bucket;

    constructor(scope: Construct, id: string) {
        super(scope, id);
        this.tmxBucket = new s3.Bucket(this, 'TmxBucket', { bucketName: 'mock-tmx-bucket-events' });
        this.documentsBucket = new s3.Bucket(this, 'DocsBucket', { bucketName: 'mock-docs-bucket-events' });
    }
}

describe('EventsConstruct', () => {
    let stack: cdk.Stack;
    let template: Template;

    beforeAll(() => {
        stack = new cdk.Stack();
        const lambdaConstruct = new MockLambdaConstruct(stack, 'MockLambda');
        const storageConstruct = new MockStorageConstruct(stack, 'MockStorage');

        new EventsConstruct(stack, 'MyTestEvents', {
            lambdaFunctions: lambdaConstruct as unknown as LambdaConstruct,
            storage: storageConstruct as unknown as StorageConstruct
        });
        template = Template.fromStack(stack);
    });

    test('Snapshot Test', () => {
        expect(template.toJSON()).toMatchSnapshot();
    });

    test('Creates Two EventBridge Rules', () => {
        template.resourceCountIs('AWS::Events::Rule', 2);
    });

    test('TMX Rule Configured Correctly', () => {
        template.hasResourceProperties('AWS::Events::Rule', {
            Description: 'Rule to trigger TMX processing lambda on S3 TMX bucket uploads',
            EventPattern: {
                source: ['aws.s3'],
                'detail-type': ['AWS API Call via CloudTrail'],
                detail: {
                    eventSource: ['s3.amazonaws.com'],
                    eventName: ['PutObject', 'CompleteMultipartUpload'],
                    requestParameters: {
                        bucketName: [{
                            Ref: Match.stringLikeRegexp('MockStorageTmxBucket.*')
                        }]
                    }
                }
            },
            State: 'ENABLED',
            Targets: Match.arrayWith([
                Match.objectLike({
                    Arn: { 'Fn::GetAtt': [Match.stringLikeRegexp('MockLambdaTmxProcessor.*'), 'Arn'] }
                })
            ])
        });
    });

    test('Document Rule Configured Correctly', () => {
        template.hasResourceProperties('AWS::Events::Rule', {
            Description: 'Rule to trigger document processing lambda on S3 documents bucket uploads',
            EventPattern: {
                source: ['aws.s3'],
                'detail-type': ['AWS API Call via CloudTrail'],
                detail: {
                    eventSource: ['s3.amazonaws.com'],
                    eventName: ['PutObject', 'CompleteMultipartUpload'],
                    requestParameters: {
                        bucketName: [{
                            Ref: Match.stringLikeRegexp('MockStorageDocsBucket.*')
                        }]
                    }
                }
            },
            State: 'ENABLED',
            Targets: Match.arrayWith([
                Match.objectLike({
                    Arn: { 'Fn::GetAtt': [Match.stringLikeRegexp('MockLambdaDocProcessor.*'), 'Arn'] }
                })
            ])
        });
    });
});