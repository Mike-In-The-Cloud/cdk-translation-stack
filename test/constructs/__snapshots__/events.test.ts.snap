// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsConstruct Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MockLambdaDocProcessor88CB8262": {
      "DependsOn": [
        "MockLambdaDocProcessorServiceRole24004A36",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2dbcc137043e884f1c386366a8bccd26f28559a85682271109e308fb2fb0aa04.zip",
        },
        "FunctionName": "dummy-DocProcessor-func",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MockLambdaDocProcessorServiceRole24004A36",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MockLambdaDocProcessorServiceRole24004A36": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MockLambdaTmxProcessor6F8A6BFB": {
      "DependsOn": [
        "MockLambdaTmxProcessorServiceRole2336FA7B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2dbcc137043e884f1c386366a8bccd26f28559a85682271109e308fb2fb0aa04.zip",
        },
        "FunctionName": "dummy-TmxProcessor-func",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MockLambdaTmxProcessorServiceRole2336FA7B",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MockLambdaTmxProcessorServiceRole2336FA7B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MockStorageDocsBucket0311B786": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "mock-docs-bucket-events",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MockStorageTmxBucket51279A9C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "mock-tmx-bucket-events",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestEventsDocumentUploadRuleAllowEventRuleMockLambdaDocProcessorBDEA2112F121C1D2": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MockLambdaDocProcessor88CB8262",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "MyTestEventsDocumentUploadRuleB48BFFDE",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestEventsDocumentUploadRuleB48BFFDE": {
      "Properties": {
        "Description": "Rule to trigger document processing lambda on S3 documents bucket uploads",
        "EventPattern": {
          "detail": {
            "eventName": [
              "PutObject",
              "CompleteMultipartUpload",
            ],
            "eventSource": [
              "s3.amazonaws.com",
            ],
            "requestParameters": {
              "bucketName": [
                {
                  "Ref": "MockStorageDocsBucket0311B786",
                },
              ],
            },
          },
          "detail-type": [
            "AWS API Call via CloudTrail",
          ],
          "source": [
            "aws.s3",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "MockLambdaDocProcessor88CB8262",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyTestEventsTMXUploadRule5B888296": {
      "Properties": {
        "Description": "Rule to trigger TMX processing lambda on S3 TMX bucket uploads",
        "EventPattern": {
          "detail": {
            "eventName": [
              "PutObject",
              "CompleteMultipartUpload",
            ],
            "eventSource": [
              "s3.amazonaws.com",
            ],
            "requestParameters": {
              "bucketName": [
                {
                  "Ref": "MockStorageTmxBucket51279A9C",
                },
              ],
            },
          },
          "detail-type": [
            "AWS API Call via CloudTrail",
          ],
          "source": [
            "aws.s3",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "MockLambdaTmxProcessor6F8A6BFB",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyTestEventsTMXUploadRuleAllowEventRuleMockLambdaTmxProcessor438161B42A02780F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MockLambdaTmxProcessor6F8A6BFB",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "MyTestEventsTMXUploadRule5B888296",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
